### PR TITLE
Support slicing StoredArrays

### PR DESCRIPTION
**Pull request by mcwitt**
Implements `__getitem__(self, slice)` for `StoredArrays`. This will be useful e.g. for subsampling frames in downstream code once #959 or #957 are merged.

### CODE REVIEW

**Review by badisa — APPROVED** [review]

**Review by jkausrelay — APPROVED** [review]

**Comment by mcwitt** [comment]
I think there is one optimization worth making: in the current implementation, the extreme case of slicing with a stride of one (e.g. `stored_arrays[:]`) loads each `.npy` file many times (equal to the number of elements in the array). Added a failing test to demonstrate this in https://github.com/proteneer/timemachine/pull/964/commits/dee102dc580a48a9510e5799329b4e6fa6c961f5.

I think a simple solution would be to use `lru_cache(maxsize=1)` to cache the last-read chunk, so traversing in order shouldn't result in multiple reads of the same file regardless of the stride. Will make this modification in an upcoming commit.

File: timemachine/fe/stored_arrays.py [hunk 1]
```diff
@@ -78,13 +82,13 @@ def __eq__(self, other) -> bool:
             np.array_equal(a, b, equal_nan=True) for a, b in zip(self, other)
         )
 
-    def _chunks(self) -> Iterator[List[NDArray]]:
+    def _chunks(self) -> Iterator[NDArray]:
```
Comment (mcwitt) [inline timemachine/fe/stored_arrays.py:85]: this type was incorrect before

**Comment by badisa** [comment]
> I think there is one optimization worth making: in the current implementation, the extreme case of slicing with a stride of one (e.g. `stored_arrays[:]`) loads each `.npy` file many times (equal to the number of elements in the array). Added a failing test to demonstrate this in [dee102d](https://github.com/proteneer/timemachine/commit/dee102dc580a48a9510e5799329b4e6fa6c961f5).
> 
> I think a simple solution would be to use `lru_cache(maxsize=1)` to cache the last-read chunk, so traversing in order shouldn't result in multiple reads of the same file regardless of the stride. Will make this modification in an upcoming commit.

Is there a way to avoid this? We currently have O(1) memory usage here, and adding an lru_cache switches it to O(N) for a case that seems like we can typically avoid in practice.

**Comment by badisa** [comment]
It seems like it should be possible to group the indices produced by the slice by their relevant chunk and doing that rather then using an LRU cache. That way we still get one read per chunk, but don't add memory overhead

**Comment by mcwitt** [comment]
> We currently have O(1) memory usage here, and adding an lru_cache switches it to O(N)

(Note that `N` here is the size of a single chunk, and most operations defined here already do use O(N) memory. To be clear, this would add an O(N) overhead for the lifetime of the object.)

> It seems like it should be possible to group the indices produced by the slice by their relevant chunk and doing that rather then using an LRU cache

This is possible, but implementing this correctly seems tricky (e.g. to support the slice interface need to handle arbitrary combinations possibly missing or negative start, end, and step).

One option would be to remove the slice capability and add a limited interface that only supports subsampling via changing the stride.

Another (maybe preferable?) option is to move the cache into the scope of the `__getitem__`. This seems like the best of both worlds.

**Comment by badisa** [comment]
> To be clear, this would add an O(N) overhead for the lifetime of the object.)

Sorry that was imprecise, yes its that it is for the lifetime of the object.

> One option would be to remove the slice capability and add a limited interface that only supports subsampling via changing the stride.

I would probably advocate for this, as it requires more explicit usage of the class. And the current implementation of the slicing in the case of `stored_arrays[:]` will store a single file per frame rather then the original of a thousand frames per file, which has its own potential pain points (in the typical case calling sample). I don't know how much value there is in having StoredArray be used generically as a `Sequence` when we have pretty specific use cases and will need to handle its usage differently already.

**Comment by mcwitt** [comment]
> And the current implementation of the slicing in the case of stored_arrays[:] will store a single file per frame rather then the original of a thousand frames per file

This is true. I'm not sure of a good way around this with the current interface. In general, it seems like supporting slicing with the current interface leads to suboptimal characteristics. Agreed that it might make more sense to deal with e.g. subsampling explicitly in downstream code.

I'm now thinking that this PR was a bit premature. Will close for now and potentially resurrect later if I can think of a better approach.

**Pull request closed without merging**